### PR TITLE
SPM-1416: Add baseline description field

### PR DESCRIPTION
**Pull request by michalslomczynski**
## Secure Coding Practices Checklist GitHub Link
- https://github.com/RedHatInsights/secure-coding-checklist

## Secure Coding Checklist
- [x] Input Validation
- [x] Output Encoding
- [x] Authentication and Password Management
- [x] Session Management
- [x] Access Control
- [x] Cryptographic Practices
- [x] Error Handling and Logging
- [x] Data Protection
- [x] Communication Security
- [x] System Configuration
- [x] Database Security
- [x] File Management
- [x] Memory Management
- [x] General Coding Practices

### CODE REVIEW

File: manager/controllers/baseline_create.go [hunk 1]
```diff
@@ -67,21 +70,20 @@ func CreateBaselineHandler(c *gin.Context) {
...
 	if err != nil {
 		LogAndRespError(c, err, "Database error")
 		return
 	}
 
 	configUpdated := request.Config != nil
-	inventoryIDs := kafka.GetInventoryIDsToEvaluate(&baselineID, accountID, configUpdated, nil)
-	kafka.EvaluateBaselineSystems(inventoryIDs)
+	inventoryIDs := kafka.GetInventoryIDsToEvaluate(&resp.BaselineID, accountID, configUpdated, nil)
+	go kafka.EvaluateBaselineSystems(inventoryIDs)
```
Comment (michalslomczynski) [inline manager/controllers/baseline_create.go:81]: @MichaelMraka what do you think about this one?

File: database_admin/schema/create_schema.sql [hunk 2]
```diff
@@ -620,6 +620,7 @@ CREATE TABLE IF NOT EXISTS baseline
     rh_account_id INT               NOT NULL REFERENCES rh_account (id),
     name          TEXT              NOT NULL,
```
Comment (MichaelMraka) [inline database_admin/schema/create_schema.sql:621]: I think there should be unique constraint on (rh_account_id, name) so customer can't have two baselines with the same name - they won't be able to distinguish them.

File: manager/controllers/baseline_create.go (hunk 1 above)
Comment (MichaelMraka) [reply manager/controllers/baseline_create.go:81]: One issue per commit. :)
If you have an idea for improvement please create an independent commit for it so it's easier do distinguish what is the primary fix/feature and what's the improvement.

File: manager/controllers/baseline_create.go [hunk 3]
```diff
@@ -16,16 +16,19 @@ import (
...
 	// Baseline config to filter applicable advisories and package updates for the associated systems (optional).
 	Config *BaselineConfig `json:"config"`
+	// Description of the baseline (optional).
+	Description string `json:"description"`
 }
 
 type CreateBaselineResponse struct {
-	BaselineID int `json:"baseline_id" example:"1"` // Updated baseline unique ID, it can not be changed
+	BaselineID  int    `json:"baseline_id" example:"1"` // Updated baseline unique ID, it can not be changed
+	Description string `json:"description,omitempty"`
```
Comment (MichaelMraka) [inline manager/controllers/baseline_create.go:31]: Do not add description to the response. UI already have this data, it put them into request creating the baseline.

**Review by MichaelMraka — CHANGES_REQUESTED** [review]

File: manager/controllers/baseline_create.go (hunk 3 above)
Comment (michalslomczynski) [reply manager/controllers/baseline_create.go:31]: Ok

File: database_admin/schema/create_schema.sql (hunk 2 above)
Comment (michalslomczynski) [reply database_admin/schema/create_schema.sql:621]: I noticed the same yesterday while testing API! Nice, I agree :slightly_smiling_face:

File: manager/controllers/baseline_create.go [hunk 4]
```diff
@@ -69,13 +72,17 @@ func CreateBaselineHandler(c *gin.Context) {
 
 	baselineID, err := buildCreateBaselineQuery(request, accountID)
 	if err != nil {
+		if strings.Contains(err.Error(), "duplicate key value") {
+			LogAndRespBadRequest(c, err, "baseline name already exists")
+			return
+		}
```
[the remark this answers is not in the document]
Comment (michalslomczynski) [reply manager/controllers/baseline_create.go:78]: Like here would be better?
https://github.com/RedHatInsights/patchman-engine/blob/3e582342bb900b92c41067a9fa2f4edfb1394810/manager/controllers/baseline_update.go#L69

File: database_admin/migrations/079_baseline_unique_name.up.sql [hunk 5]
```diff
@@ -0,0 +1 @@
+ALTER TABLE baseline ADD CONSTRAINT baseline_name_rh_account_id_key UNIQUE(name, rh_account_id);
```
Comment (michalslomczynski) [inline database_admin/migrations/079_baseline_unique_name.up.sql:1]: Should I merge them both to 78, or we are fine?

Comment (michalslomczynski) [reply database_admin/migrations/079_baseline_unique_name.up.sql]: I merged

File: manager/controllers/baseline_create.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (michalslomczynski) [reply manager/controllers/baseline_create.go]: I think I did that correctly now.

File: manager/controllers/baseline_detail.go [hunk 6]
```diff
@@ -15,7 +15,8 @@ import (
 )
 
 type BaselineDetailResponse struct {
-	Data BaselineDetailItem `json:"data"`
+	Data        BaselineDetailItem `json:"data"`
+	Description string             `json:"description,omitempty"`
```
Comment (MichaelMraka) [inline manager/controllers/baseline_detail.go:19]: Should not be `Description` part of `Data`, actually part of `BaselineDetailAttributes` type?

File: manager/controllers/baseline_update.go [hunk 7]
```diff
@@ -89,22 +93,22 @@ func BaselineUpdateHandler(c *gin.Context) {
...
+		switch e := err.(type) {
+		case *pgconn.PgError:
+			if e.Code == errDuplicatedKeyVal {
+				LogAndRespBadRequest(c, err, "baseline name already exists")
+				return
+			}
+		default:
+			LogAndRespError(c, err, "Database error")
+			return
+		}
```
Comment (MichaelMraka) [inline manager/controllers/baseline_update.go:105]: This is looks like slightly modified copy-paste of `baseline_create.go:76`.
Duplicated code mostly means it should be a function (somewhere in database/utils in this case) called on two places ;-).

**Review by MichaelMraka — CHANGES_REQUESTED** [review]

File: manager/controllers/baseline_update.go (hunk 7 above)
Comment (michalslomczynski) [reply manager/controllers/baseline_update.go:105]: Could you propose what you have on your mind?

File: manager/controllers/baseline_detail.go (hunk 6 above)
Comment (michalslomczynski) [reply manager/controllers/baseline_detail.go:19]: Probably yes

File: manager/controllers/baseline_create.go [hunk 8]
```diff
@@ -9,19 +9,23 @@ import (
...
-
-	"github.com/gin-gonic/gin"
 )
 
+const errDuplicatedKeyVal = "23505"
+
 type CreateBaselineRequest struct {
-	// Baseline name (optional).
+	// Baseline name.
 	Name *string `json:"name"`
```
[the remark this answers is not in the document]
Comment (michalslomczynski) [reply manager/controllers/baseline_create.go:22]: I dont know here what Pepa had on his mind, dont ask me please. I can put this comment back.

[the remark this answers is not in the document]
Comment (michalslomczynski) [reply manager/controllers/baseline_create.go:22]: Furthermore it is not part of this feature.

File: manager/controllers/baseline_update.go (hunk 7 above)
Comment (michalslomczynski) [reply manager/controllers/baseline_update.go:105]: I just noticed however potential bug there when error is of type pgconn.PgError and code is different. Will fix that.

File: manager/controllers/baseline_detail.go (hunk 6 above)
Comment (michalslomczynski) [reply manager/controllers/baseline_detail.go:19]: I changed that. Looks much better now :slightly_smiling_face: Thanks for catching this.

**Review by MichaelMraka — APPROVED** [review]
I've put some improvements into PR.

**Pull request merged**